### DISCUSSION TITLE
Looping Approach

### DISCUSSION THREAD

**Discussion by rhyswynn** [Holmeswww/AgentKit#18] [Q&A] [answered]
Hi there!  When building and managing a graph for a dynamic process that might have to iterate through some nodes multiple times, would the approach be to build the graph and begin the evaluation with only the nodes defined up to the first branch\loop, and then build the rest of the graph dynamically? Would the logic at that branching point happen external to the graph and nodes, or internal, like within an after query? Like maybe use a separate graph for the loop and evaluate the results, and then pass those results into a new graph with new nodes for the final output? A single db node could be used throughout. It seems like it would be easier to manage the branching and looping 'outside' of the graph elements. Thank you!

**Comment by Holmeswww** [comment, marked as answer]
I think both external/internal can do the looping.

The benefit of internal is that the nodes in the loop will have direct access to prior computation through `dependencies`.
The benefit of external is that it is easier to maintain.

The idea of using a separate graph for the loop is also interesting.

**Reply by rhyswynn** [reply]
Thank you for the response! I'm looking through the code some more, and it seems like a true looping process is not possible. The graph evaluates nodes in sequence based on the edge and order rules, and a node that has already been executed is added to the history and excluded from being re-queued. So it would seem there could be no true loops, either it would be a long chain of temporary nodes with unique keys, or multiple instances of separate, short graph chains. Is my understanding correct?

**Reply by Holmeswww** [reply]
Your understanding is correct. There can be no loops in the graph. But you can keep adding a node to implement a loop.

**Reply by rhyswynn** [reply]
I created some test functions to just pass the values around and tested adding the temporary nodes and edges. Now I understand how that works. It doesn't help to think of it as looping, it's more like repeating the same temporary short branches. Thank you so much for your help and patience!
